soroban-simulation/src/snapshot_source.rs

```diff
@@ -90,70 +123,105 @@ impl<T: SnapshotSourceWithArchive> AutoRestoringSnapshotSource<T> {
-                        }
-                    };
+        let entry_state = self.snapshot_source.get_including_archived(key)?;
+        match entry_state.state {
+            LedgerEntryArchivalState::Live => (),
+            LedgerEntryArchivalState::New(need_proof) => {
+                if need_proof {
+                    let mut new_keys_requiring_proof = self
+                        .new_keys_requiring_proof
+                        .try_borrow_mut()
+                        .map_err(|_| {
+                            HostError::from((ScErrorType::Context, ScErrorCode::InternalError))
+                        })?;
+                    new_keys_requiring_proof.insert(key.clone());
+                }
+            }
+            LedgerEntryArchivalState::Archived(_) => {
+                if entry_state.entry.is_some() {
+                    let mut restored_ledger_keys =
+                        self.restored_ledger_keys.try_borrow_mut().map_err(|_| {
+                            HostError::from((ScErrorType::Context, ScErrorCode::InternalError))
+                        })?;
+                    restored_ledger_keys.insert(key.clone());
+                }
+            }
+        }
+        if let Some(entry) = entry_state.entry {
+            if let Some(_durability) = get_key_durability(key.as_ref()) {
+                if let Some(live_until) = entry_state.live_until_ledger {
+                    if live_until < self.current_ledger_sequence {
+                        return Err(HostError::from((
+                            ScErrorType::Storage,
+                            ScErrorCode::InternalError,
+                        )));
+                    }
+                } else {
+                    if !matches!(entry_state.state, LedgerEntryArchivalState::Archived(_)) {
+                        return Err(HostError::from((
+                            ScErrorType::Storage,
+                            ScErrorCode::InternalError,
+                        )));
+                    }
+                    return Ok(Some((entry, Some(self.min_persistent_live_until_ledger))));
+                }
+            } else {
+                if !matches!(entry_state.state, LedgerEntryArchivalState::Live)
+                    && !matches!(entry_state.state, LedgerEntryArchivalState::New(false))
+                {
+                    // Something went wrong, entries without durability can't be archived or
+                    // require a proof.
+                    return Err(HostError::from((
+                        ScErrorType::Storage,
+                        ScErrorCode::InternalError,
+                    )));
                 }
             }
-            Ok(Some((entry, live_until)))
+            Ok(Some((entry, entry_state.live_until_ledger)))
         } else {
-            Ok(None)
+            if matches!(entry_state.state, LedgerEntryArchivalState::Archived(_)) {
```

Comment:
Nit: This invariant can be strengthened to `if !matches!(entry_state.state, LedgerEntryArchivalState::New(_)`